keps/sig-api-machinery/3962-mutating-admission-policies/README.md

```diff
@@ -0,0 +1,1363 @@
+  The number of types to check are also heavily limited to prevent the checks taking up too much computing time from the KCM.
+
+#### Composition variables
+To control the size of mutation expression, and to better reuse parts of the expressions, sub-expressions can be extracted into a separate variables section, similar to variables of ValidatingAdmissionPolicy.
+```yaml
+variables:
+  - name: targetContainers
+    expression: >-
+      object.spec.template.containers.filter(c,
+      c.image.contains("example.com"))
+  - name: transformedContainers
+    expression: >
+      variables.targetContainers.map(c, {"name": c.name, "env": {"name": "FOO",
+      "value": "foo"}})
+mutations:
+  - mutation: |
+      Object{
+          spec: Object.spec{
+              template: Object.spec.template{
+                  containers: variables.transformedContainers
+              }
+          }
+      }
+
+```
+
+With variable composition, it is possible to escape from compile-time type checking. For example
+```yaml
+variables:
+  - name: definitelyNotAContainer # resulting type is Dyn
+    expression: >-
+      params.foo == "bar" ? true : "true"
+mutations:
+    - mutation: |
+      Object{
+          spec: Object.spec{
+              template: Object.spec.template{
+                  containers: [variables.definitelyNotAContainer] # will pass, but error at runtime.
+              }
+          }
+      }
+
+```
+
+### Risk
+
+1. Ensure the final state match expectation.
+There might be multiple mutating admission policies, mutating webhooks, other controllers trying to mutate the incoming request and each happens separately, 
+and they might mutate the same part of the object. 
+It might be hard to ensure that the final state matches expectations. 
+For best practice, the validation process is highly recommended whenever there is a mutation process set up. The validating admission policy is recommended to be set up whenever a mutation admission policy is set to verify the final state of the data matches the expectation. Also refer to the Safety section for further details.
+
+2. Failures in MutatingAdmissionPolicy will fail request in admission chain.
+If the failure policy is set to fail and the mutation admission policy matches all resources, the failure/error in MAP might infect the control plane availability.
+
+3. Interference other process.
+If the mutation admission policy tries to mutate the pod resources during create/update, it might interfere with other processes like replicaSet.
+
+### User Stories
+
+#### Use case: Set a label
+
+```cel
+Object.spec{
+  Object.metadata{
+    labels:
+      "label-to-set": "label-value"
+  }
+}
+```
+
+#### Use case: Set a field
+
+```cel
+Object{
+    spec: Object.spec{
+        containers: object.spec.containers.map(c,
+            Object.spec.containers.item{
+                name: c.name,
```

Comment:
does this set the image pull policy in a single container or in all containers?

Reply:
This is for all containers with map macro :)